


<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="shortcut icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-5.5.0">
    
    
      
        <title>Minotaor</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.b5d04df8.min.css">
      
        <link rel="stylesheet" href="assets/stylesheets/palette.9ab2c1f8.min.css">
      
      
        
        
        <meta name="theme-color" content="#4caf50">
      
    
    
    
      
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
    
    
      
    
    
  </head>
  
  
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="green" data-md-color-accent="lightgreen">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#minotaor" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href="." title="Minotaor" class="md-header-nav__button md-logo" aria-label="Minotaor">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      
        <div class="md-header-nav__ellipsis">
          <span class="md-header-nav__topic md-ellipsis">
            Minotaor
          </span>
          <span class="md-header-nav__topic md-ellipsis">
            
              Home
            
          </span>
        </div>
      
    </div>
    
      <label class="md-header-nav__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active">
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" data-md-component="search-reset" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/Edinburgh-Genome-Foundry/Minotaor/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    Minotaor
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
        
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="Minotaor" class="md-nav__button md-logo" aria-label="Minotaor">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    Minotaor
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/Edinburgh-Genome-Foundry/Minotaor/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    Minotaor
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        Home
        <span class="md-nav__icon md-icon">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 9h14V7H3v2m0 4h14v-2H3v2m0 4h14v-2H3v2m16 0h2v-2h-2v2m0-10v2h2V7h-2m0 6h2v-2h-2v2z"/></svg>
        </span>
      </label>
    
    <a href="." title="Home" class="md-nav__link md-nav__link--active">
      Home
    </a>
    
      
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#background" class="md-nav__link">
    Background
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#install" class="md-nav__link">
    Install
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage" class="md-nav__link">
    Usage
  </a>
  
    <nav class="md-nav" aria-label="Usage">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#plotting" class="md-nav__link">
    Plotting
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#reference-datasets" class="md-nav__link">
    Reference datasets
  </a>
  
    <nav class="md-nav" aria-label="Reference datasets">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#immune-epitope-database" class="md-nav__link">
    Immune Epitope Database
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#vdjdb" class="md-nav__link">
    VDJdb
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prosite" class="md-nav__link">
    PROSITE
  </a>
  
    <nav class="md-nav" aria-label="PROSITE">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#conversion" class="md-nav__link">
    Conversion
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#versioning" class="md-nav__link">
    Versioning
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#license-mit" class="md-nav__link">
    License = MIT
  </a>
  
</li>
      
    </ul>
  
</nav>
    
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-nav__toggle md-toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2">
    
    <label class="md-nav__link" for="nav-2">
      Reference
      <span class="md-nav__icon md-icon">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M8.59 16.58L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.42z"/></svg>
      </span>
    </label>
    <nav class="md-nav" aria-label="Reference" data-md-level="1">
      <label class="md-nav__title" for="nav-2">
        <span class="md-nav__icon md-icon">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
        </span>
        Reference
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-nav__toggle md-toggle" data-md-toggle="nav-2-1" type="checkbox" id="nav-2-1">
    
    <label class="md-nav__link" for="nav-2-1">
      Minotaor
      <span class="md-nav__icon md-icon">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M8.59 16.58L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.42z"/></svg>
      </span>
    </label>
    <nav class="md-nav" aria-label="Minotaor" data-md-level="2">
      <label class="md-nav__title" for="nav-2-1">
        <span class="md-nav__icon md-icon">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
        </span>
        Minotaor
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="reference/minotaor/" title="Index" class="md-nav__link">
      Index
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="reference/minotaor/minotaor/" title="Minotaor" class="md-nav__link">
      Minotaor
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="reference/minotaor/version/" title="Version" class="md-nav__link">
      Version
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#background" class="md-nav__link">
    Background
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#install" class="md-nav__link">
    Install
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage" class="md-nav__link">
    Usage
  </a>
  
    <nav class="md-nav" aria-label="Usage">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#plotting" class="md-nav__link">
    Plotting
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#reference-datasets" class="md-nav__link">
    Reference datasets
  </a>
  
    <nav class="md-nav" aria-label="Reference datasets">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#immune-epitope-database" class="md-nav__link">
    Immune Epitope Database
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#vdjdb" class="md-nav__link">
    VDJdb
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prosite" class="md-nav__link">
    PROSITE
  </a>
  
    <nav class="md-nav" aria-label="PROSITE">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#conversion" class="md-nav__link">
    Conversion
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#versioning" class="md-nav__link">
    Versioning
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#license-mit" class="md-nav__link">
    License = MIT
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/Edinburgh-Genome-Foundry/Minotaor/edit/master/README.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                  
                
                
                <p align="center">
<img alt="Minotaor logo" title="Minotaor" src="https://raw.githubusercontent.com/Edinburgh-Genome-Foundry/Minotaor/main/images/minotaor.png" width="120">
</p>

<h1 id="minotaor">Minotaor</h1>
<p><a href="https://travis-ci.org/Edinburgh-Genome-Foundry/Minotaor"><img alt="Build Status" src="https://travis-ci.org/Edinburgh-Genome-Foundry/Minotaor.svg?branch=main" /></a>
<a href="https://coveralls.io/github/Edinburgh-Genome-Foundry/Minotaor?branch=main"><img alt="Coverage Status" src="https://coveralls.io/repos/github/Edinburgh-Genome-Foundry/Minotaor/badge.svg?branch=main" /></a></p>
<p><strong>Work in progress</strong></p>
<p>Minotaor is an a<strong>mino</strong> acid sequence anno<strong>ta</strong>t<strong>or</strong> for quickly identifying common protein tags and linkers in an ORF. Additionally, it can flag peptide motifs that are known to cause problems during translation. It uses Biopython.</p>
<h2 id="background">Background</h2>
<p>In the <a href="https://prosite.expasy.org/">PROSITE</a> nomenclature, a sequence <em>motif</em> is a description
of the occurrence of amino acids (signature, fingerprint), and can be either a pattern or a profile.
A <em>pattern</em> is a qualitative description of a motif in a regular expression-like syntax.
A <em>profile</em> (or weight matrix) is a table of position-specific amino acid weights and gap costs.</p>
<h2 id="install">Install</h2>
<div class="codehilite"><pre><span></span><code><span class="err">pip install minotaor</span>
</code></pre></div>


<h2 id="usage">Usage</h2>
<div class="codehilite"><pre><span></span><code><span class="kn">import</span> <span class="nn">minotaor</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>

<span class="n">protein</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;SYYHHHHHHDYDIPTTENLYFQG*EDINBURGHGENQMEFQUNDRY*&quot;</span><span class="p">)</span>
<span class="n">protein_record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">protein</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;example&quot;</span><span class="p">,</span> <span class="n">annotations</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;molecule_type&quot;</span><span class="p">:</span> <span class="s2">&quot;protein&quot;</span><span class="p">})</span>

<span class="n">protein_record</span> <span class="o">=</span> <span class="n">minotaor</span><span class="o">.</span><span class="n">annotate_record</span><span class="p">(</span><span class="n">protein_record</span><span class="p">)</span>  <span class="c1"># search is case sensitive</span>
</code></pre></div>


<h3 id="plotting">Plotting</h3>
<p>Plotting requires <a href="https://github.com/Edinburgh-Genome-Foundry/DnaFeaturesViewer">DNA Features Viewer</a> installed:</p>
<div class="codehilite"><pre><span></span><code><span class="n">graphic_record</span> <span class="o">=</span> <span class="n">minotaor</span><span class="o">.</span><span class="n">MinotaorTranslator</span><span class="p">()</span><span class="o">.</span><span class="n">translate_record</span><span class="p">(</span><span class="n">protein_record</span><span class="p">)</span>
<span class="n">ax</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">graphic_record</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figure_width</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">strand_in_label_threshold</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
<span class="n">graphic_record</span><span class="o">.</span><span class="n">plot_sequence</span><span class="p">(</span><span class="n">ax</span><span class="p">)</span>
</code></pre></div>


<p><img alt="Example" src="images/example.png" /></p>
<h3 id="reference-datasets">Reference datasets</h3>
<p>Minotaor can use custom reference pandas dataframes, specified with <code>seq_dataset</code>. The <code>sequence</code> and <code>name</code> columns are used for search and naming of the motifs. The <code>type</code> column sets the search type, which can be <code>seq</code> for strings or <code>pattern</code> or regexes (regular expressions). This is shown in the below examples of epitope datasets.</p>
<h4 id="immune-epitope-database">Immune Epitope Database</h4>
<p>Download and unzip a <em>CSV Metric Export</em> of your choice from the <a href="https://www.iedb.org/database_export_v3.php">IEDB website</a>, then:</p>
<div class="codehilite"><pre><span></span><code><span class="kn">import</span> <span class="nn">pandas</span>
<span class="n">iedb_file</span> <span class="o">=</span> <span class="s1">&#39;epitope_full_v3.csv&#39;</span>
<span class="n">iedb</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">iedb_file</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>
<span class="n">iedb</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Epitope IRI&quot;</span><span class="p">:</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="s2">&quot;sequence&quot;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">iedb</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;seq&#39;</span>
<span class="c1"># The dataframe can be used as shown above (note: this is a huge dataset):</span>
<span class="n">protein_record</span> <span class="o">=</span> <span class="n">minotaor</span><span class="o">.</span><span class="n">annotate_record</span><span class="p">(</span><span class="n">protein_record</span><span class="p">,</span> <span class="n">seq_dataset</span><span class="o">=</span><span class="n">iedb</span><span class="p">)</span>
</code></pre></div>


<p>The epitopes can then be looked up with the link provided as their names. Alternatively,
a more informative epitope name can be constructed from the other columns, as shown in the next section.</p>
<h4 id="vdjdb">VDJdb</h4>
<p>Download and unzip the latest <a href="https://github.com/antigenomics/vdjdb-db/releases/latest">VDJdb release</a>, then:</p>
<div class="codehilite"><pre><span></span><code><span class="kn">import</span> <span class="nn">pandas</span>
<span class="n">vdjdb_file</span> <span class="o">=</span> <span class="s1">&#39;vdjdb-YYYY-MM-DD/vdjdb.slim.txt&#39;</span>
<span class="n">vdjdb</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">vdjdb_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="c1"># Create a unique subset of the epitopes:</span>
<span class="n">vdjdb_dataset</span> <span class="o">=</span> <span class="n">vdjdb</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">vdjdb_dataset</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;antigen.gene&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">vdjdb_dataset</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">vdjdb_dataset</span><span class="p">[</span><span class="s1">&#39;antigen.epitope&#39;</span><span class="p">]</span>  <span class="c1"># or &#39;cdr3&#39; for antibodies</span>
<span class="n">vdjdb_dataset</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;seq&#39;</span>
<span class="n">vdjdb_dataset</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;VDJdb epitope &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">antigen</span><span class="p">)</span>
                         <span class="k">for</span> <span class="n">antigen</span> <span class="ow">in</span> <span class="n">vdjdb_dataset</span><span class="p">[</span><span class="s1">&#39;antigen.gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()]</span>

<span class="n">protein_record</span> <span class="o">=</span> <span class="n">minotaor</span><span class="o">.</span><span class="n">annotate_record</span><span class="p">(</span><span class="n">protein_record</span><span class="p">,</span> <span class="n">seq_dataset</span><span class="o">=</span><span class="n">vdjdb_dataset</span><span class="p">)</span>
</code></pre></div>


<p>The motifs then can be looked up in VDJdb for more details.
A similar approach can be used for the McPAS-TCR database.</p>
<h3 id="prosite">PROSITE</h3>
<p>A function is provided for reading <a href="https://prosite.expasy.org/scanprosite">ScanProsite</a> results.
Query your sequence with Biopython:</p>
<div class="codehilite"><pre><span></span><code><span class="kn">from</span> <span class="nn">Bio.ExPASy</span> <span class="kn">import</span> <span class="n">ScanProsite</span>
<span class="n">my_seq</span> <span class="o">=</span> <span class="s1">&#39;MYHHHHHHYAGDLPGLMDGAAAGGGA&#39;</span>  <span class="c1"># cannot contain &#39;*&#39;</span>
<span class="n">scanprosite_handle</span> <span class="o">=</span> <span class="n">ScanProsite</span><span class="o">.</span><span class="n">scan</span><span class="p">(</span><span class="n">seq</span><span class="o">=</span><span class="n">my_seq</span><span class="p">,</span> <span class="n">mirror</span><span class="o">=</span><span class="s1">&#39;https://prosite.expasy.org/&#39;</span><span class="p">,</span> <span class="n">output</span><span class="o">=</span><span class="s1">&#39;xml&#39;</span><span class="p">)</span>
<span class="n">scanprosite_record</span> <span class="o">=</span> <span class="n">ScanProsite</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">scanprosite_handle</span><span class="p">)</span>

<span class="n">protein_record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">my_seq</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;my_seq&quot;</span><span class="p">,</span> <span class="n">annotations</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;molecule_type&quot;</span><span class="p">:</span> <span class="s2">&quot;protein&quot;</span><span class="p">})</span>
<span class="n">protein_record</span> <span class="o">=</span> <span class="n">minotaor</span><span class="o">.</span><span class="n">add_scanprosite_results</span><span class="p">(</span><span class="n">protein_record</span><span class="p">,</span> <span class="n">scanprosite_record</span><span class="p">)</span>
</code></pre></div>


<p><img alt="Prosite" src="images/example_prosite.png" /></p>
<h4 id="conversion">Conversion</h4>
<p>Convert between PROSITE and regex formats:</p>
<div class="codehilite"><pre><span></span><code><span class="n">regex</span> <span class="o">=</span> <span class="n">minotaor</span><span class="o">.</span><span class="n">convert_prosite_to_regex</span><span class="p">(</span><span class="s2">&quot;&lt;A-[GV]-</span><span class="si">{PR}</span><span class="s2">-[FYW](2)-</span><span class="si">{P}</span><span class="s2">(4)-x-x(8)&gt;.&quot;</span><span class="p">)</span>
<span class="n">regex</span>
<span class="c1"># &#39;^A[GV][^PR][FYW]{2}[^P]{4}[^\\*][^\\*]{8}$&#39;</span>
<span class="n">minotaor</span><span class="o">.</span><span class="n">convert_regex_to_prosite</span><span class="p">(</span><span class="n">regex</span><span class="p">)</span>
<span class="c1"># &#39;&lt;A-[GV]-{PR}-[FYW](2)-{P}(4)-x-x(8)&gt;.&#39;</span>
</code></pre></div>


<p>Given a DNA sequence, return amino acid sequences that may contain it:</p>
<div class="codehilite"><pre><span></span><code><span class="n">bsmbi_site</span> <span class="o">=</span> <span class="s2">&quot;CGTCTC&quot;</span>
<span class="nb">print</span><span class="p">(</span><span class="n">minotaor</span><span class="o">.</span><span class="n">convert_dna_to_aa_pattern</span><span class="p">(</span><span class="n">bsmbi_site</span><span class="p">))</span>
<span class="c1"># [&#39;RL&#39;, &#39;[DATRSICYNLFPHVG]V[S]&#39;, &#39;[SPAT]S[RPLQH]&#39;, &#39;ET&#39;, &#39;[MAT*RSPKLEVQGW]R[R]&#39;, &#39;[G*R]D[DAVEG]&#39;]</span>
</code></pre></div>


<p>Returns a regex for each of the 6 translation frames. Note that the search with regex is case sensitive.</p>
<h2 id="versioning">Versioning</h2>
<p>Minotaor uses the <a href="https://semver.org">semantic versioning</a> scheme.</p>
<h2 id="license-mit">License = MIT</h2>
<p>Minotaor is <a href="https://www.gnu.org/philosophy/free-sw.en.html">free software</a>, which means
the users have the freedom to run, copy, distribute, study, change and improve the software.</p>
<p>Minotaor was written at the <a href="https://edinburgh-genome-foundry.github.io/">Edinburgh Genome Foundry</a>
by <a href="https://github.com/veghp">Peter Vegh</a> and is released under the MIT license.</p>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
        
          <a href="reference/minotaor/" title="Index" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Index
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Powered by
        <a href="http://timothycrosley.github.io/portray">portray.</a>
        You too can
        <a href="http://timothycrosley.github.io/portray">
          portray</a>
        your Python project well using automatic documentation.
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="assets/javascripts/vendor.92ffa368.min.js"></script>
      <script src="assets/javascripts/bundle.5123e3d4.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents"}</script>
      
      <script>
        app = initialize({
          base: ".",
          features: [],
          search: Object.assign({
            worker: "assets/javascripts/worker/search.a68abb33.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>